jujugui/static/gui/src/app/state/test-state.js

```diff
@@ -711,31 +711,228 @@ describe('State', () => {
     });
   });
 
+  describe('State.register()', () => {
+    it('stores the supplied dispatchers', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123',
+        seriesList:  ['precise', 'trusty', 'xenial']
+      });
+      const stub1 = sinon.stub();
+      const stub2 = sinon.stub();
+      const stub3 = sinon.stub();
+      const stub4 = sinon.stub();
+      const stub5 = sinon.stub();
+      state.register([
+        ['*', stub1],
+        ['*', stub5],
+        ['store', stub2],
+        // dispatchers with cleanup argument.
+        ['machine', stub3, stub4]
+      ]);
+      assert.deepEqual(state._dispatchers, {
+        '*': [[stub1, undefined], [stub5, undefined]],
+        store: [[stub2, undefined]],
+        machine: [[stub3, stub4]]
+      });
+    });
+  });
+
   describe('State.dispatch()', () => {
     it('passes the current location to generateState', () => {
       const state = new window.jujugui.State({
         baseURL: 'http://abc.com:123',
         seriesList:  ['precise', 'trusty', 'xenial'],
-        location: {href: 'foo'}
+        location: {href: '/hatch/ghost'}
       });
       const stub = sinon.stub(
         state, 'generateState', () => ({ error: null, state: {}}));
       state.dispatch();
       assert.equal(stub.callCount, 1);
-      assert.deepEqual(stub.args[0], ['foo']);
+      assert.deepEqual(stub.args[0], ['/hatch/ghost']);
     });
 
     it('updates the _appStateHistory with the new state', () => {
       const state = new window.jujugui.State({
         baseURL: 'http://abc.com:123',
         seriesList:  ['precise', 'trusty', 'xenial'],
-        location: {href: 'foo'}
+        location: {href: 'hatch/ghost'}
       });
       sinon.stub(state,
         'generateState', () => ({ error: null, state: {new: 'state'}}));
       state.dispatch();
       assert.deepEqual(state._appStateHistory, [{new: 'state'}]);
     });
+
+    it('dispatches registered dispatchers in proper order', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123',
+        seriesList:  ['precise', 'trusty', 'xenial'],
+        location: {href: 'ghost/trusty/i/machines'}
+      });
+      let counter = 0;
+      let increment = () => counter += 1;
+      let execution = {};
+      const stub = function(state, next) {
+        execution.stub = increment();
+        next();
+      };
+      const stub2 = function(state, next) {
+        execution.stub2 = increment();
+        next();
+      };
+      const stub3 = function(state, next) {
+        execution.stub3 = increment();
+        next();
+      };
+      const stub4 = function(state, next) {
+        execution.stub4 = increment();
+        next();
+      };
+      state.register([
+        ['*', stub],
+        ['store', stub2],
+        ['*', stub3],
+        ['gui.machines', stub4]
+      ]);
+      state.dispatch();
+      assert.deepEqual(execution, {stub: 1, stub2: 4, stub3: 2, stub4: 3});
```

Comment:
Also, can we safely assume that stub4 is executed before stub2 in this case? Aren't these taken from the dictionary? Is there an order?